### PR TITLE
Add Exporters to Prometheus

### LINKED ISSUE #41: [FEATURE] Add Exporters to Prometheus
**Is your feature request related to a problem? Please describe.**
add exporters in teler, so we can get a statistical data 

**Describe the solution you'd like**
N/A

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by JustHumanz**
### Summary
- Add Exporters to Prometheus for monitoring in realtime and easy to analyze the log (#41)

### Proposed of changes

This PR fixes/implements the following **bugs/features**:
- Add Exporters to Prometheus

### How has this been tested?
Add -m/--metrics parameter for `Run as exporter`
Proof:
example query in prometheus

### Closing issues

Fixes #41 

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have followed the guidelines in our [CONTRIBUTING.md](https://github.com/kitabisa/teler/blob/development/.github/CONTRIBUTING.md) document.
- [ ] I have written new tests for my changes.
  - [ ] My changes successfully ran and pass tests locally.

### CODE REVIEW

**Comment by dwisiswant0** [comment]
Can you please rebase it again from `master`, @JustHumanz?

File: README.md [hunk 1]
```diff
@@ -332,6 +334,15 @@ alert:
   active: true
   provider: "slack"
 ```
+### Collected metrics
+| Name                        | Desc                            |
+|-----------------------------|---------------------------------|
+|`teler_threats_count_total`  | the total amount of threats     |
```
Comment (dwisiswant0) [inline README.md:340]: It should be: `Total number of detected threats`.

File: README.md [hunk 2]
```diff
@@ -332,6 +334,15 @@ alert:
...
+|`teler_cwa`                  | get all payload from `Common Web Attack` threats  |
```
Comment (dwisiswant0) [inline README.md:341]: It should be: `Get lists of Common Web Attacks`.

File: README.md [hunk 3]
```diff
@@ -332,6 +334,15 @@ alert:
...
+|`teler_badcrawler`           | list useragent and remote address of threats              |
```
Comment (dwisiswant0) [inline README.md:342]: It should be: `Get lists of Bad Crawler requests`.

File: README.md [hunk 4]
```diff
@@ -332,6 +334,15 @@ alert:
...
+|`teler_dir_bruteforce`       | list of bruteforced dir by threats          |
```
Comment (dwisiswant0) [inline README.md:343]: It should be: `Get lists of Directories Bruteforced`.

File: README.md [hunk 5]
```diff
@@ -332,6 +334,15 @@ alert:
...
+|`teler_badip_count`          | count of bad ip address       |
```
Comment (dwisiswant0) [inline README.md:344]: It should be: `Total number of Bad IP Addresses`.

File: README.md [hunk 6]
```diff
@@ -332,6 +334,15 @@ alert:
...
+|`teler_bad_referrer`         | http request with a bad referrer          |
```
Comment (dwisiswant0) [inline README.md:345]: It should be: `Get lists of Bad Referrer requests`.

File: internal/runner/runner.go [hunk 7]
```diff
@@ -27,6 +28,10 @@ func New(options *common.Options) {
 	var wg sync.WaitGroup
 	var input *os.File
 
+	if options.Metrics {
+		log.Infof("Listening metrics on port : " + strconv.Itoa(options.MetricsPort))
```
Comment (dwisiswant0) [inline internal/runner/runner.go:32]: Please remove trailing spaces for this.
```diff
-		log.Infof("Listening metrics on port : " + strconv.Itoa(options.MetricsPort))
+		log.Infof("Listening metrics on port: " + strconv.Itoa(options.MetricsPort))
```

File: pkg/teler/metrics.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
+package teler
+
+import "github.com/prometheus/client_golang/prometheus"
+
+var (
+	getcwa = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_cwa",
+			Help: "teler_cwa show all cwa threats",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:9]: It should be: `Get lists of Common Web Attack threats`.

File: pkg/teler/metrics.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		},
+		[]string{"description", "remote_addr", "request_uri", "status"},
+	)
+
+	getbadcrawler = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_badcrawler",
+			Help: "teler_badcrawler get a bad crawler",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:17]: It should be: `Get lists of Bad Crawler requests`.

File: pkg/teler/metrics.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+		},
+		[]string{"remote_addr", "http_user_agent", "status"},
+	)
+
+	getdirbruteforce = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_dir_bruteforce",
+			Help: "teler_dir_bruteforce list of bruteforced dir",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:25]: It should be: `Get lists of Directories Bruteforced`.

File: pkg/teler/metrics.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+			Name: "teler_bad_referrer",
+			Help: "get a http with bad referrer",
+		},
+		[]string{"http_referer"},
+	)
+
+	getthreatstotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_threats_count_total",
+			Help: "get threats count",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:49]: It should be: `Total number of detected threats`.

File: pkg/teler/metrics.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+			Name: "teler_dir_bruteforce",
+			Help: "teler_dir_bruteforce list of bruteforced dir",
+		},
+		[]string{"remote_addr", "request_uri", "status"},
+	)
+
+	getbadip = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_badip_count",
+			Help: "get_badip remote address of threats",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:33]: It should be: `Total number of Bad IP Addresses`.

File: pkg/teler/metrics.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+		},
+		[]string{"remote_addr"},
+	)
+
+	getbadreferrer = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_bad_referrer",
+			Help: "get a http with bad referrer",
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:41]: It should be: `Get lists of Bad Referrer requests`.

File: cmd/teler/main.go [hunk 14]
```diff
@@ -14,5 +18,16 @@ func init() {
...
 	options := runner.ParseOptions()
-	runner.New(options)
+	if options.Metrics {
+		http.Handle("/metrics", promhttp.Handler())
+
+		teler.MetricInit()
+		go runner.New(options)
+
+		http.ListenAndServe(":"+strconv.Itoa(options.MetricsPort), nil)
+
```
Comment (dwisiswant0) [inline cmd/teler/main.go:28]: Please remove trailing newline in this.

File: cmd/teler/main.go [hunk 15]
```diff
@@ -14,5 +18,16 @@ func init() {
...
+	} else {
+		runner.New(options)
+	}
+
```
Comment (dwisiswant0) [inline cmd/teler/main.go:32]: Please remove trailing newline in this.

File: pkg/teler/metrics.go [hunk 16]
```diff
@@ -0,0 +1,60 @@
+package teler
+
+import "github.com/prometheus/client_golang/prometheus"
+
+var (
+	getcwa = prometheus.NewCounterVec(
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:6]: We use camelCase for naming convention, please change this to `getCWA`.

File: pkg/teler/metrics.go [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+		prometheus.CounterOpts{
+			Name: "teler_cwa",
+			Help: "teler_cwa show all cwa threats",
+		},
+		[]string{"description", "remote_addr", "request_uri", "status"},
+	)
+
+	getbadcrawler = prometheus.NewCounterVec(
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:14]: We use camelCase for naming convention, please change this to `getBadCrawler`.

File: pkg/teler/metrics.go [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+		prometheus.CounterOpts{
+			Name: "teler_badcrawler",
+			Help: "teler_badcrawler get a bad crawler",
+		},
+		[]string{"remote_addr", "http_user_agent", "status"},
+	)
+
+	getdirbruteforce = prometheus.NewCounterVec(
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:22]: We use camelCase for naming convention, please change this to `getDirBruteforce`.

File: pkg/teler/metrics.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+	getbadip = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "teler_badip_count",
+			Help: "get_badip remote address of threats",
+		},
+		[]string{"remote_addr"},
+	)
+
+	getbadreferrer = prometheus.NewCounterVec(
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:38]: We use camelCase for naming convention, please change this to `getBadReferrer`.

File: pkg/teler/metrics.go [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+		prometheus.CounterOpts{
+			Name: "teler_bad_referrer",
+			Help: "get a http with bad referrer",
+		},
+		[]string{"http_referer"},
+	)
+
+	getthreatstotal = prometheus.NewCounterVec(
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:46]: We use camelCase for naming convention, please change this to `getThreatsTotal`.

File: pkg/teler/metrics.go [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+		prometheus.CounterOpts{
+			Name: "teler_threats_count_total",
+			Help: "get threats count",
+		},
+		[]string{"case"},
+	)
+)
+
+//add var to register
```
Comment (dwisiswant0) [inline pkg/teler/metrics.go:55]:
```diff
-//add var to register
+// MetricInit will register a metric with the specified variables
```

**Review by dwisiswant0 — CHANGES_REQUESTED** [review]
When I try to run it with command as:
```bash
> [buffers] | ./bin/teler -m -p
```
Program doesn't run port with the specified default value.

I see `-m` metric flag does nothing but mediate true/false logic.
It's better to be refactored as: `-m <PORT>`, if the `PORT` value isn't provided, then we run metric with the default value we specify.

I also saw you create a new function that was exported as `MetricInit`, it's also better if you create that function as a new package _(e.g. `metrics`)_, and not in `teler` package itself.

If you have rebased again, please revise the review as needed.
Thank you!

**Comment by JustHumanz** [comment]
wha,thx for suggestion,gonna fix it

**Comment by dwisiswant0** [comment]
Looks like you haven't resolved everything I've asked for changes. Please re-check carefully, @JustHumanz.

**Comment by JustHumanz** [comment]
> Looks like you haven't resolved everything I've asked for changes. Please re-check carefully, @JustHumanz.

all done,can you testing it?
i'm not very sure with my code ._.

**Review by dwisiswant0 — APPROVED** [review]

**Comment by dwisiswant0** [comment]
Since I need a little rework from this, I'm going to change the base branch to `development`.
I really appreciate your efforts to contribute to teler development, @JustHumanz. Accepting now!

**Comment by dwisiswant0** [comment]
@all-contributors please add @JustHumanz for code. :man_dancing:

**Pull request merged**